deep_qa/data/vocabulary.py

```diff
@@ -7,35 +7,38 @@
 
 logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
 
+class _NamespaceDependentDefaultDict(defaultdict):
+    def __init__(self, non_padded_namespaces: List[str], padded_function, non_padded_function):
+        self._non_padded_namespaces = non_padded_namespaces
+        self._padded_function = padded_function
+        self._non_padded_function = non_padded_function
+        super(_NamespaceDependentDefaultDict, self).__init__()
+
+    def __missing__(self, key: str):
+        value = None
+        for namespace_str in self._non_padded_namespaces:
+            if namespace_str[0] == '*' and key.endswith(namespace_str[1:]):
+                value = self._non_padded_function()
+            elif namespace_str == key:
+                value = self._non_padded_function()
+        if value is None:
+            value = self._padded_function
+        dict.__setitem__(self, key, value)
```

Comment:
Yes, it is definitely good practice to avoid these.  I think I know the reason to avoid it here, though - you'll run into a recursive loop.  `self[key]` calls `__missing__` if `key not in self`.